clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2583,11 +2580,27 @@ struct ConvertConstructorToDeductionGuideTransform {
 
     //    -- The types of the function parameters are those of the constructor.
     for (auto *OldParam : TL.getParams()) {
-      ParmVarDecl *NewParam =
-          transformFunctionTypeParam(OldParam, Args, MaterializedTypedefs);
-      if (NestedPattern && NewParam)
+      ParmVarDecl *NewParam = OldParam;
+      // Given
+      //   template <class T> struct C {
+      //     template <class U> struct D {
+      //       template <class V> D(U, V);
+      //     };
+      //   };
+      // First, transform all the references to template parameters that are
+      // defined outside of the surrounding class template. That is T in the
+      // above example.
+      if (NestedPattern) {
         NewParam = transformFunctionTypeParam(NewParam, OuterInstantiationArgs,
                                               MaterializedTypedefs);
+        if (!NewParam)
+          return QualType();
+      }
+      // Then, transform all the references to template parameters that are
+      // defined at the class template and the constructor. In this example,
+      // they're U and V, respectively.
+      NewParam =
+          transformFunctionTypeParam(NewParam, Args, MaterializedTypedefs);
```

Comment:
We have the same pattern on Line 2438-2452 in this file, where we perform a substitution of `OuterInstantiationArgs` on a new transformed parameter decl, I think we should probably fix it as well.

Reply:
I think that's just fine, although it looks weird: those lines are merely "creating" new template parameters rather than doing type substitution. (it's applying a `TemplateDeclInstantiator` with the instantiating arguments - the only valuable part from these arguments is the depth we're supposed to deduct, and the type isn't something we're concerned about at that time.)

Frankly, I have to say I didn't fully understand why we have such discrepancies (use the `transformTemplateParameter` first and then a `TemplateDeclInstantiator` for nested cases). I was expecting that perhaps we could unify them in one call. I think it warrants a separate refactoring PR, then.

Reply:
The `TemplateDeclInstantiator` is used for that substitution to avoid evaluating constraints at that point. Without this, the constraint template argument depths on `NewParam` would be as if the parameter were at depth 0 (instead of the nested template depth), causing a crash when evaluated later.

Reply:
@antangelo : thanks for the explanation. I was considering if we can reuse the `transformTemplateTypeParam` stuff instead of a `TemplateDeclInstantiator` (we can make the constraint evaluation opt-in in that function; I presumed this would behave like what `setEvaluateConstraints(false)` does.), but doing that doesn't look like it would simplify the logic here significantly, so I hesitated to do so.

The other part `Depth1Args` also prevents us from combining these substitution into one call - I think I have to go through it further. Anyhow, this shouldn't block a backport.